Add start and stop callbacks

jevin opened the pull request:
This PR adds lifecycle events to Sinatra. Related issue: #1184

jevin commented:
This is still missing tests. I'm not sure where they fit it, anyone can give me some pointers?

jevin commented:
Okay this is ready for a review! @zzak (sorry I'm not sure who else to @')

dentarg commented:
The code looks good to me.

This is only used if you use Sinatra to start you web server, is that correct?

Can you add some documentation to the README? Maybe even add an example of this to the `examples` directory?

Thanks

jevin commented:
Hey @dentarg, sorry I missed your comment!

> This is only used if you use Sinatra to start you web server, is that correct?

Correct!

> Can you add some documentation to the README? Maybe even add an example of this to the `examples` directory?

I will get on that!

jevin commented:
@dentarg I just updated the README and added `lifecycle_events.rb` in `examples`!

Also, if you feel like the lifecycle events would make sense for servers NOT started with Sinatra, I'm down to create another PR for this!

dentarg commented on lib/sinatra/base.rb:
```diff
@@ -1559,6 +1567,8 @@ def quit!
         set :running_server, nil
         set :handler_name, nil
+
+        on_stop_callback.call if !on_stop_callback.nil?
```
```suggestion
        on_stop_callback.call unless on_stop_callback.nil?
```

dentarg commented on lib/sinatra/base.rb:
```diff
@@ -1646,7 +1656,7 @@ def start_server(handler, server_settings, handler_name)
           set :handler_name,   handler_name
           server.threaded = settings.threaded if server.respond_to? :threaded=
-
+          on_start_callback.call if !on_start_callback.nil?
```
```suggestion
          on_start_callback.call unless on_start_callback.nil?
```

dentarg reviewed (commented):
I think this reads easier

dentarg commented on test/server_test.rb:
```diff
@@ -27,6 +27,14 @@ def stop
   register 'mock', 'Rack::Handler::Mock'
 end
 
+class EventHookTest
+  def self.start
+  end
+
+  def self.stop
+  end
+end
```
I think you can make this an anonymous class only available inside `ServerTest`: https://dev.to/factorial/a-trick-with-ruby-anonymous-classes-11pp

jevin replied:
Sure thing! I've turned it into an anonymous class.

jevin commented:
@dentarg, I've implemented the changes. Have a look and let me know!

dentarg commented:
Great, do you mind rebasing to get CI green?
